### PR TITLE
Add authentication wrapper function and readme

### PR DESCRIPTION
**Pull request by CarlHembrough**
### What

All existing API's are using the same function to wrap endpoints with an auth check. I have updated the function to work with identities stored in request context and added it to go-ns for reuse.

Added a readme to document the usage of the middleware and wrapper function.

### How to review

Review changes

### Who can review

Anyone

### CODE REVIEW

File: identity/authentication.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+		callerIdentity := r.Context().Value(callerIdentityKey)
+
+		// just checking if an identity exists until permissions are being provided.
+		if callerIdentity == nil || callerIdentity == ""{
+			http.Error(w, "requested resource not found", http.StatusNotFound)
+			log.Error(errors.New("client missing auth token in header"), nil)
+			return
+		}
+
+		// The request has been authenticated, now run the clients request
```
Comment (daiLlew) [inline identity/authentication.go]: Is it worth logging that the request has been authenticated?

Comment (CarlHembrough) [reply identity/authentication.go]: Added a log statement at the start of the handler and at the end when its complete

File: identity/authentication_test.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+	Convey("Given a context with an identity", t, func() {
+
+		ctx := context.WithValue(context.Background(), callerIdentityKey, "user@ons.gov.uk")
+
+		Convey("When IsPresent is called with the context", func() {
+
+			identityIsPresent := IsPresent(ctx)
+
+			Convey("Then a 404 response is returned", func() {
```
Comment (gedge) [inline identity/authentication_test.go]: incorrect statement or test

Comment (CarlHembrough) [reply identity/authentication_test.go]: Copy paste fail - I have updated the statement 👍

**Pull request merged**